apps/ensindexer/src/config/app-config.ts

```diff
@@ -0,0 +1,97 @@
+import { ENSIndexerConfigSchema } from "@/config/config.schema";
+import { ENSIndexerConfig, ENSIndexerEnvironment, RawChainConfig } from "@/config/types";
+import { z } from "zod/v4";
+
+/**
+ * Extracts dynamic chain configuration from environment variables.
+ *
+ * This function scans all environment variables for keys matching the pattern
+ * "RPC_URL_{chainId}", where {chainId} must be a string of digits (e.g., "1", "10", "8453").
+ *
+ * It then looks for matching environment variables for the rate limit of requests per second
+ * for the given chain, using the pattern "RPC_REQUEST_RATE_LIMIT_{chainId}".
+ *
+ * This function returns a raw chain config which is not yet validated against the zod schema.
+ */
+export function getChainsFromEnv(): Record<number, RawChainConfig> {
+  const chains: Record<number, RawChainConfig> = {};
+
+  Object.entries(process.env).forEach(([key, value]) => {
+    // Only match keys like "RPC_URL_1", "RPC_URL_10", etc. (digits only after the underscore)
+    const match = key.match(/^RPC_URL_(\d+)$/);
+
+    // If the key after `RPC_URL_` is not a number or the value is empty, skip
+    if (!match || !value) return;
+
+    // The regex above ensures that only numeric chain IDs are matched.
+    // - Example: "RPC_URL_1" will match and extract "1" as the chainId.
+    // - Example: "RPC_URL_SOMESTRING" will NOT match, so no risk of NaN from non-numeric IDs.
+    const chainId = Number(match[1]);
+
+    // Optionally get the rate limit for this chain, if set.
+    const rpcMaxRequestsPerSecond = process.env[`RPC_REQUEST_RATE_LIMIT_${chainId}`];
+
+    chains[chainId] = {
+      // The value for each RPC_URL_{chainId} is used as the rpcEndpointUrl.
+      rpcEndpointUrl: value,
+      // Note: rpcMaxRequestsPerSecond is typed as string!!!
+      // Type coercion is handled later by zod validation in the actual config.
+      rpcMaxRequestsPerSecond: rpcMaxRequestsPerSecond,
+    };
+  });
+
+  return chains;
+}
+
+function getValidationErrors(issues: z.core.$ZodIssue[]) {
+  // loop over and concatenate the messages. precede with a line break due
+  // to the way ponder prints errors
+  return `\n` + issues.map((issue) => issue.message).join("\n");
+}
+
+// loads the relevant environment variables in the shape of the zod schema
+function parseEnvironment(): ENSIndexerEnvironment {
+  const config: ENSIndexerEnvironment = {
```

Comment:
nit: just `return { ... }` should suffice with the return type specified in the function, or perhaps `return { ... } satisfies ENSIndexerEnvironment` to be explicit